### PR TITLE
doc: explain how to write ooni nettests

### PR DESCRIPTION
**Pull request by bassosimone**
## Checklist

- [x] I have read the [contribution guidelines](https://github.com/ooni/probe-cli/blob/master/CONTRIBUTING.md)
- [x] reference issue for this pull request: https://github.com/ooni/ooni.org/issues/361 and https://github.com/ooni/probe/issues/1733.
- [x] related ooni/spec pull request: N/A

Location of the issue tracker: https://github.com/ooni/probe

## Description

This is a preliminary commit and requires more work. I am going
to push this commit so others soon can start to review it.

This draft supersedes https://github.com/ooni/probe-cli/pull/505

This diff also contains a library of building blocks that is
instrumental to complete websteps. In fact, the tutorial I have
written here will eventually guide people to write their own
implementation of websteps.

I have not written any test for the new code so far. It is
also a stretch to call this new code. It is existing code that
has been written by me or @kelmenhorst that I have adapted
for the purpose of writing the tutorial.

The nice exercise of writing a tutorial is that you are more
easily frustrated if the code is difficult to use XD.

## How to review the tutorials

Navigate to `internal/tutorial/measure` and follow the links
inside the `README.md` file. Or go to https://github.com/ooni/probe-cli/tree/WORK/internal/tutorial/measure
and read starting from there, which is equivalent.

Editing is quite tricky. Most text comes from comments in
Go files. You need to edit the comments and then you need
to regenerate all the files. The `README.md` of the
`internal/tutorial/measure` tutorial contains instructions
on how to do that.

## What remains to be done before merging

Mainly adding tests and making sure I didn't lose
anything when adapting existing code. I like the interface
of the new code more, but existing code written last
August may have a better implementation of corner cases.

### CODE REVIEW

File: internal/measure/quic.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		ctx, "udp", address, tlsConfig, &quic.Config{})
+	m.Completed = time.Since(qh.begin)
+	if err != nil {
+		m.Failure = err
+		return m
+	}
+	select {
+	case <-sess.HandshakeComplete().Done():
+	case <-ctx.Done():
+		m.Failure = err
```
Comment (bassosimone) [inline internal/measure/quic.go:73]: TODO(@bassosimone): this line is wrong.
```suggestion
		m.Failure = ctx.Err()
```

File: internal/tutorial/measure/chapter07/main.go [hunk 2]
```diff
@@ -0,0 +1,833 @@
...
+	// and we make an instance of the `urlMeasurer`. This type
+	// represents the possibility of measuring an URL (without any
+	// redirection for now).
+	//
+	// When we initialize the `urlMeasurer`, we pass to it an
+	// instance of `Measurer` created like we did in previous chapters.
+	//
+	// ```Go
+	begin := time.Now()
+	trace := measure.NewTrace(begin)
```
Comment (bassosimone) [inline internal/tutorial/measure/chapter07/main.go:70]: TODO(bassosimone): using the same trace for all endpoints is wrong.

Comment (bassosimone) [reply internal/tutorial/measure/chapter07/main.go:70]: Maybe it would make the code more robust to assign an `int64` ID to every connection, so that we can assert in the (future) tests that the network events of a given endpoint only contains a single connection.

**Comment by bassosimone** [comment]
I flagged a couple of places that were wrong. After re-reading, it seems to me all chapters could benefit from more subsections with titles. This will allow to provide comments before and after snippets of code with a more clear delimiter.

File: internal/tutorial/measure/chapter07/main.go (hunk 2 above)
Comment (bassosimone) [reply internal/tutorial/measure/chapter07/main.go]: Done by refactoring and also created a TODO comment for the ID for every conn.

File: internal/measure/quic.go (hunk 1 above)
Comment (bassosimone) [reply internal/measure/quic.go]: Fixed.

**Comment by bassosimone** [comment]
Superseded by https://github.com/ooni/probe-cli/pull/506
